documentation/directory-structure.md

```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
 > The structure described in this article is strictly enforced. There exist some exceptions for historical reasons.
 > These exceptions are not allowed going forward.
 
+## Key concepts
+
+The directory structure is a reflection of few key concepts.
+
+A `service` is a set of operation endpoints (typically HTTP REST API endpoints) that **version uniformly**.
+
+A `service group` is a set of services that share common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
+
+In case of ARM, a `service` consists of multiple `resource types`.
+A `resource type` can be derived from the URL of HTTP API operations that pertain to it.
+
+### Uniform versioning
+
+A `service` **must version uniformly**. This means the following:
+
+1. Any deployed service operation endpoint must belong to an API version. An API version once deployed is immutable.
+   Its behavior or constituent operations cannot be changed.
+1. Any given service API version can be composed only of HTTP API operations and ARM resource types (if applicable)
+   that have the same API version. 
+1. Any documentation pertaining to the service and any SDKs generated from the service must pertain to only
+   one service version.
+1. The service version must be always represented in its entirety; any SDK or documentation referring to the service
+   must encompass all of it.
+1. The `common-types` OpenAPI definition shared across multiple services can version independently of the service.
+   However, the rule that API version is immutable still must be observed.
+
+The **uniform versioning** has several implications and implementation decisions supporting it:
+
+- The service is effectively defined as a series of consecutive API versions.
+- Each API version is represented by a pair of folders representing its lifecycle stage and release date, 
```

Comment:
It's not actually a release date. It is just a date. It _could_ be a release date, or it could be the date when the new version started being implemented. The only requirement is that each date me later in time than the previous date - the date has no actual meaning.

Reply:
One other thing: it is illegal to have both stable and preview versions with the same date. For example, 2024-05-08-preview and 2024-05-08 should NEVER happen. The reason is because it is impossible for the customer to tell if the stable came after the preview or vice verse.  Some teams might try to go from preview to stable by simply truncating the `-preview` suffix; this is NOT acceptable. Instead, the team would have to pick a later date for the stable which could be just 1 day later: 2024-05-09